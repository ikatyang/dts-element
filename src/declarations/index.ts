// tslint:disable:max-line-length

export * from './class-declaration';
export * from './enum-declaration';
export * from './function-declaration';
export * from './generic-declaration';
export * from './global-declaration';
export * from './interface-declaration';
export * from './module-declaration';
export * from './namespace-declaration';
export * from './parameter-declaration';
export * from './type-declaration';
export * from './variable-declaration';
