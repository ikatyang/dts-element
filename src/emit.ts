import * as ts from 'typescript';
import { IElement } from './element';
import { transform } from './transform';

export const emit_native = (node: ts.Node, options?: ts.PrinterOptions) => {
  const source_file = ts.createSourceFile(
    '',
    '',
    ts.ScriptTarget.Latest,
    false,
    ts.ScriptKind.TS,
  );
  const printer = ts.createPrinter(options);
  return printer.printNode(ts.EmitHint.Unspecified, node, source_file);
};

export const emit = (element: IElement<any>, options?: ts.PrinterOptions) =>
  emit_native(transform(element), options);
