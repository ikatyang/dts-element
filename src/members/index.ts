export * from './class-member';
export * from './export-member';
export * from './import-member';
export * from './object-member';
