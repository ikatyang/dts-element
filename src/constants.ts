import * as ts from 'typescript';
import { create_native_type, INativeType } from './types/native-type';

export enum ElementKind {
  ArrayType,
  ClassDeclaration,
  ClassMember,
  ConditionalType,
  ConstructorType,
  EnumDeclaration,
  ExportDefault,
  ExportEqual,
  ExportMember,
  ExportNamed,
  ExportNamespace,
  FunctionDeclaration,
  FunctionType,
  GeneralType,
  GenericDeclaration,
  GlobalDeclaration,
  ImportEqual,
  ImportMember,
  ImportNamed,
  ImportNamespace,
  IndexSignature,
  InferType,
  InterfaceDeclaration,
  IntersectionType,
  JSDocComment, // tslint:disable-line:naming-convention
  KeyofType,
  LiteralType,
  MappedType,
  ModuleDeclaration,
  MultiLineComment,
  NamespaceDeclaration,
  NativeType,
  ObjectMember,
  ObjectType,
  ParameterDeclaration,
  SingleLineComment,
  SubType,
  TopLevelElement,
  TripleSlashReference,
  TupleType,
  TypeDeclaration,
  TypePredicate,
  TypeofType,
  UnionType,
  VariableDeclaration,
}

export const any_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword),
});
export const boolean_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.BooleanKeyword),
});
export const never_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.NeverKeyword),
});
export const null_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.NullKeyword),
});
export const number_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.NumberKeyword),
});
export const object_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.ObjectKeyword),
});
export const string_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.StringKeyword),
});
export const symbol_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.SymbolKeyword),
});
export const this_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.ThisKeyword),
});
export const undefined_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.UndefinedKeyword),
});
export const void_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.VoidKeyword),
});
