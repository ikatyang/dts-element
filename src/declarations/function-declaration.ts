import * as ts from 'typescript';
import { ElementKind } from '../constants';
import { create_element, IElement, IElementOptions } from '../element';
import { transform } from '../transform';
import { create_function_type, IFunctionType } from '../types/function-type';
import { add_declare_modifier_if_need } from '../utils';

export interface IFunctionDeclarationOptions extends IElementOptions {
  name: string | undefined;
  type?: IFunctionType;
  export?: boolean;
}

export interface IFunctionDeclaration
  extends IElement<ElementKind.FunctionDeclaration>,
    IFunctionDeclarationOptions {}

export const create_function_declaration = (
  options: IFunctionDeclarationOptions,
): IFunctionDeclaration => ({
  ...create_element(ElementKind.FunctionDeclaration),
  ...options,
});

// tslint:disable:ter-indent

/**
 * @hidden
 */
export const transform_function_declaration = (
  element: IFunctionDeclaration,
  path: IElement<any>[],
) => {
  const function_type = transform(
    element.type || create_function_type(),
    path,
  ) as ts.FunctionDeclaration;
  return ts.createFunctionDeclaration(
    /* decorators      */ undefined,
    /* modifiers       */ add_declare_modifier_if_need(
      element.export === true
        ? [ts.createToken(ts.SyntaxKind.ExportKeyword)]
        : undefined,
      path,
    ),
    /* asteriskToken   */ undefined,
    /* name            */ element.name,
    /* typeParameters  */ function_type.typeParameters,
    /* parameters      */ function_type.parameters,
    /* type            */ function_type.type,
    /* body            */ undefined,
  );
};
