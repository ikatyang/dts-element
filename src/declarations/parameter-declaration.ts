import * as ts from 'typescript';
import { IType } from '../collections';
import { any_type } from '../constants';
import { ElementKind } from '../constants';
import {
  create_element,
  is_element,
  IElement,
  IElementOptions,
} from '../element';
import { transform } from '../transform';
import { create_array_type } from '../types/array-type';

export interface IParameterDeclarationOptions extends IElementOptions {
  name: string;
  type?: IType;
  rest?: boolean;
  optional?: boolean;
}

export interface IParameterDeclaration
  extends IElement<ElementKind.ParameterDeclaration>,
    IParameterDeclarationOptions {}

export const create_parameter_declaration = (
  options: IParameterDeclarationOptions,
): IParameterDeclaration => ({
  ...create_element(ElementKind.ParameterDeclaration),
  ...options,
});

export const is_parameter_declaration = (
  value: any,
): value is IParameterDeclaration =>
  is_element(value) && value.kind === ElementKind.ParameterDeclaration;

// tslint:disable:ter-indent

/**
 * @hidden
 */
export const transform_parameter_declaration = (
  element: IParameterDeclaration,
  path: IElement<any>[],
) =>
  ts.createParameter(
    /* decorators      */ undefined,
    /* modifiers       */ undefined,
    /* dotDotDotToken  */ element.rest === true
      ? ts.createToken(ts.SyntaxKind.DotDotDotToken)
      : undefined,
    /* name            */ element.name,
    /* questionToken   */ element.optional === true
      ? ts.createToken(ts.SyntaxKind.QuestionToken)
      : undefined,
    /* type            */ transform(
      element.rest === true
        ? create_array_type({ type: element.type || any_type })
        : element.type || any_type,
      path,
    ) as ts.TypeNode,
    /* initializer     */ undefined,
  );
