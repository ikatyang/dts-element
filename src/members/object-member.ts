import * as ts from 'typescript';
import { any_type, ElementKind } from '../constants';
import { IFunctionDeclaration } from '../declarations/function-declaration';
import { IVariableDeclaration } from '../declarations/variable-declaration';
import {
  create_element,
  is_element,
  IElement,
  IElementOptions,
} from '../element';
import { transform } from '../transform';
import { IConstructorType } from '../types/constructor-type';
import { create_function_type } from '../types/function-type';
import { is_valid_identifier, mutable_array } from '../utils';

export interface IObjectMemberOptions extends IElementOptions {
  owned: IVariableDeclaration | IFunctionDeclaration | IConstructorType;
  optional?: boolean;
  readonly?: boolean;
}

export interface IObjectMember
  extends IElement<ElementKind.ObjectMember>,
    IObjectMemberOptions {}

export const create_object_member = (
  options: IObjectMemberOptions,
): IObjectMember => ({
  ...create_element(ElementKind.ObjectMember),
  ...options,
});

export const is_object_member = (value: any): value is IObjectMember =>
  is_element(value) && value.kind === ElementKind.ObjectMember;

/**
 * @hidden
 */
export const transform_object_member = (
  element: IObjectMember,
  path: IElement<any>[],
) => {
  const modifiers =
    element.readonly === true
      ? [ts.createToken(ts.SyntaxKind.ReadonlyKeyword)]
      : undefined;
  const question_token =
    element.optional === true
      ? ts.createToken(ts.SyntaxKind.QuestionToken)
      : undefined;
  switch (element.owned.kind) {
    case ElementKind.VariableDeclaration:
      return ts.createPropertySignature(
        /* modifiers     */ modifiers,
        /* name          */ typeof element.owned.name === 'string' &&
        is_valid_identifier(element.owned.name)
          ? element.owned.name
          : (ts.createLiteral(element.owned.name) as
              | ts.StringLiteral
              | ts.NumericLiteral),
        /* questionToken */ question_token,
        /* type          */ transform(
          element.owned.type || any_type,
          path,
        ) as ts.TypeNode,
        /* initializer   */ undefined,
      );
    case ElementKind.FunctionDeclaration:
      const function_type = transform(
        element.owned.type || create_function_type(),
        path,
      ) as ts.FunctionDeclaration;
      return element.owned.name === undefined
        ? ts.createCallSignature(
            /* typeParameters  */ mutable_array(function_type.typeParameters),
            /* parameters      */ mutable_array(function_type.parameters),
            /* type            */ function_type.type,
          )
        : ts.createMethodSignature(
            /* typeParameters  */ function_type.typeParameters,
            /* parameters      */ function_type.parameters,
            /* type            */ function_type.type,
            /* name            */ is_valid_identifier(element.owned.name)
              ? element.owned.name
              : ts.createLiteral(element.owned.name),
            /* questionToken   */ question_token,
          );
    case ElementKind.ConstructorType:
      const constructor_type = transform(
        element.owned,
        path,
      ) as ts.ConstructorTypeNode;
      return ts.createConstructSignature(
        /* typeParameters  */ mutable_array(constructor_type.typeParameters),
        /* parameters      */ mutable_array(constructor_type.parameters),
        /* type            */ constructor_type.type,
      );
    default:
      throw new Error(
        `Unexpected kind ${ElementKind[element.kind]} ( ${element.kind} )`,
      );
  }
};
