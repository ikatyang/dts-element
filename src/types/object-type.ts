import * as ts from 'typescript';
import { ElementKind } from '../constants';
import {
  create_element,
  is_element,
  IElement,
  IElementOptions,
} from '../element';
import { IObjectMember } from '../members/object-member';
import { IIndexSignature } from '../others/index-signature';
import { transform } from '../transform';

export interface IObjectTypeOptions extends IElementOptions {
  members?: (IObjectMember | IIndexSignature)[];
}

export interface IObjectType
  extends IElement<ElementKind.ObjectType>,
    IObjectTypeOptions {}

export const create_object_type = (
  options: IObjectTypeOptions = {},
): IObjectType => ({
  ...create_element(ElementKind.ObjectType),
  ...options,
});

export const is_object_type = (value: any): value is IObjectType =>
  is_element(value) && value.kind === ElementKind.ObjectType;

/**
 * @hidden
 */
export const transform_object_type = (
  element: IObjectType,
  path: IElement<any>[],
) =>
  ts.createTypeLiteralNode(
    /* members  */ (element.members || [])
      .map(member => transform(member, path) as ts.TypeElement),
  );
