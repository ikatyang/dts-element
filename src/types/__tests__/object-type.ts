import { string_type } from '../../constants';
import { create_variable_declaration } from '../../declarations';
import { create_function_declaration } from '../../declarations/function-declaration';
import { create_parameter_declaration } from '../../declarations/parameter-declaration';
import { emit } from '../../emit';
import { create_object_member } from '../../members/object-member';
import { create_index_signature } from '../../others/index-signature';
import { create_object_type, is_object_type } from '../object-type';

it('should return correctly', () => {
  expect(
    emit(
      create_object_type({
        members: [
          create_object_member({
            owned: create_function_declaration({
              name: 'fn',
            }),
          }),
          create_index_signature({
            parameter: create_parameter_declaration({
              name: 'key',
              type: string_type,
            }),
          }),
        ],
      }),
    ),
  ).toMatchSnapshot();
});

it('should return correctly with numerical name', () => {
  expect(
    emit(
      create_object_type({
        members: [
          create_object_member({
            owned: create_variable_declaration({
              name: 0,
            }),
          }),
        ],
      }),
    ),
  ).toMatchSnapshot();
});

describe('is_object_type', () => {
  it('should return correctly', () => {
    const element = create_object_type({
      members: [],
    });
    expect(is_object_type(element)).toBe(true);
  });
});
