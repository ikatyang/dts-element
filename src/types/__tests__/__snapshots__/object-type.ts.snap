// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correctly 1`] = `
"{
    fn(): any;
    [key: string]: any;
}"
`;

exports[`should return correctly with numerical name 1`] = `
"{
    0: any;
}"
`;
